---
title: "Melanoma Study: miRNAs and other variables predicting therapy response"
output:
  html_notebook:
    theme: united
    toc: yes
author: Marc Bender
---

<style>
p.caption {
  font-size: 0.8em;
}
body {
text-align: justify}
</style>


# Preamble

This document serves as a summary of the data analysis for the melanoma study using blood from patients to investigate several parameters, e.g. miRNA expression, staging, BRAF, etc. The basic question was to investigate differences in serum markers and its correlation to immunotherapy outcome. In a final step a predictive model should be developed to use certain markers to predict therapy outcome. All miRNAs in the following study were human miRNAs. Data is available at:

  + Fireplex Analysis: https://raw.githubusercontent.com/MBender1992/MelanomaStudy/Marc/Data/miRNA_Expression_Fireplex_Melanoma_Study.csv
  + Metadata: https://raw.githubusercontent.com/MBender1992/MelanomaStudy/Marc/Data/Metadata_Melanoma_Study.csv

### Loading of required packages

```{r, message = FALSE}
# load packages "add publications of the respective packages"

# data wrangling
library(tidyverse)
library(data.table)
library(DescTools)
library(pbapply)
library(missForest)
# data visualization
library(ggpubr)
# statistical analysis
library(rstatix)
# summary table
library(table1)
# modelling/machine learning
library(caret) 
library(pROC)
library(cvAUC)
# necessary to load data from github with "source_url"
library(devtools)

# load custom written functions
source_url("https://raw.githubusercontent.com/MBender1992/base_scripts/Marc/R_functions.R")  
```


# 1. Patient characteristics

miRNA expression was measured with the Fireplex Immunoassay (Abcam) and normalized on the 12 most stable miRNAs (?). Other parameters were assessed in the clinical workflow of patient treatment. To combine miRNA expression results and patient metadata in a reproducible way, data was ingested using the following function 

```{r, message = FALSE}
load_melanoma_data <- function(){
  require(devtools)
  
  url_miR <- "https://raw.githubusercontent.com/MBender1992/MelanomaStudy/Marc/Data/miRNA_Expression_Fireplex_Melanoma_Study.csv" 
  url_meta <- "https://raw.githubusercontent.com/MBender1992/MelanomaStudy/Marc/Data/Metadata_Melanoma_Study.csv" 
  
  
  # load csv files
  dat_miR   <- read_csv(url(url_miR)) 
  dat_meta  <- read_csv(url(url_meta)) %>%
    select(-c(therapy_start, Abnahmedatum)) %>%
    mutate(TRIM_PDL1_Expression = str_replace_all(TRIM_PDL1_Expression,"\\++","+")) %>% 
    mutate(TRIM_PDL1_Expression = ifelse(TRIM_PDL1_Expression == "o", NA,TRIM_PDL1_Expression)) %>%
    mutate(Stadium = toupper(Stadium)) %>%
    mutate(Stadium = str_extract(Stadium, "^[IV]{1,3}")) %>%
    mutate(BRAF = str_replace_all(BRAF, "\\.", "")) %>% 
    mutate(breslow_thickness_mm = parse_number(breslow_thickness_mm))
  
    # change ID column to uniform capital letters for later filtering
  names(dat_miR) <- c("miRNA", toupper(names(dat_miR)[-1]))
  
  # define IDs to be dropped for further analyses
  controls <- c("K104_1", "K104_2", "K104_3A", "K104_3B")
  duplicates <- c("22B","38B","39B","47B")
  
  # wide miR data (78 patients with miRNA data)
  dat_miR_trans <- transpose_dataframe(colnames = c("ID",dat_miR$miRNA), data = dat_miR) %>%
    filter(!ID %in% controls & !ID %in% duplicates) %>%   #drop duplicate patient data 
    mutate(ID = parse_number(ID)) #convert ID to numeric
  
  # join both tables
  right_join(dat_miR_trans,dat_meta, by="ID") %>% 
    filter(!ID %in% c(1,2)) %>% # no data available for patient 1 and 2 but still part of the source table
    mutate(miRExpAssess = ifelse(is.na(rowSums(.[,which(str_detect(names(.),"mir"))])), 0,1))  %>%# if no miRNA expression has been measured fill in 0
    arrange(ID) %>% 
    mutate(prior_BRAF_therapy = ifelse(str_detect(Vorbehandlung,"Mek|Dabra|Tafinlar|Tefinlar|MEK|BRAF|Vemu|[zZ]ellboraf"), 1, 0)) %>%
    select(-Vorbehandlung)
}

dat <- load_melanoma_data() %>% 
  mutate(Responder =  factor(Responder, levels = c("nein", "ja") , labels = c("no", "yes")))
dat
```

To gain insights of the different patients a summary table was generated showing several different features split by immunotherapy response (Table 1). 


```{r, message = FALSE}
dat_table1 <- dat
setDT(dat_table1)

# define which factors to display in table
dat_table1$sex <- factor(dat_table1$sex, levels = c("m", "w") , labels = c("Male", "Female"))
dat_table1$miRExpAssess <- factor(dat_table1$miRExpAssess, levels = c(0, 1) , labels = c("no", "yes"))
dat_table1$Responder <- factor(dat_table1$Responder, levels = c("no", "yes",2) , labels = c("no", "yes","P-value"))
dat_table1$adjuvant_IFN <- factor(dat_table1$adjuvant_IFN, levels = c("nein", "ja") , labels = c("no", "yes"))
dat_table1$brainMet <- factor(dat_table1$brainMet, levels = c("nein", "ja") , labels = c("no", "yes"))
dat_table1$subtype <- factor(dat_table1$subtype, levels = c("cutanes Melanom", "Schleimhautmelanom") , labels = c("cutaneous", "mucosal"))
dat_table1$ECOG <- factor(dat_table1$ECOG, levels = c(0,1,2) , labels = c("0", "1", "2"))
dat_table1$Stadium <- factor(dat_table1$Stadium, levels = c("II", "III","IV") , labels = c("II", "III","IV"))
dat_table1$prior_BRAF_therapy <- factor(dat_table1$prior_BRAF_therapy, levels = c(0, 1) , labels = c("no", "yes"))


# define labels for the table
label(dat_table1$Alter)      <- "Age (years)"
label(dat_table1$BRAF)      <- "BRAF-status"
label(dat_table1$Stadium)  <- "AJCC stage" # add Stadium to source table
label(dat_table1$therapy_at_blood_draw) <- "Therapy at blood draw"
label(dat_table1$sex)  <- "Sex"
label(dat_table1$Responder)  <- "Immunotherapy response"
label(dat_table1$ECOG)      <- "ECOG"
label(dat_table1$breslow_thickness_mm)      <- "Breslow thickness (mm)" # change to double
label(dat_table1$subtype) <- "Subtype"
label(dat_table1$localization) <- "Localization"
label(dat_table1$brainMet) <- "Brain metastasis"
label(dat_table1$miRExpAssess) <- "miRNA expression measured"
label(dat_table1$adjuvant_IFN) <- "Received adjuvant IFN treatment"
label(dat_table1$prior_BRAF_therapy) <- "Received prior anti-BRAF therapy"


# function to display p-values  
rndr <- function(x, name, ...) {
  if (length(x) == 0) {
    y <- dat_table1[[name]] 
    ind <- !is.na(y)
    y <- y[ind]
    s <- rep("", length(render.default(x=y, name=name, ...)))
    if (is.numeric(y)) {
      p <- t.test(y ~ dat_table1$Responder[ind])$p.value
    } else {
      p <- chisq.test(table(y, droplevels(dat_table1$Responder[ind])))$p.value
    }
    s[2] <- sub("<", "&lt;", format.pval(p, digits=3, eps=0.001))
    s
  } else {
    render.default(x=x, name=name, ...)
  }
}

rndr.strat <- function(label, n, ...) {
  ifelse(n==0, label, render.strat.default(label, n, ...))
}

# define text for footnote
fn <- "Statistical test: Unequal variance t-test (welch's t-test) for numerical data and chi² test for categorical data. Raw p-values are shown."

table1(~ Alter + BRAF + prior_BRAF_therapy + Stadium + miRExpAssess + adjuvant_IFN + brainMet + sex + ECOG + breslow_thickness_mm + subtype + localization | Responder,
       data=dat_table1, droplevels=F, render=rndr, render.strat=rndr.strat, footnote = fn)

```


# 2. Association of different parameters with therapy response

## 2.1 Serum markers

```{r, message = FALSE}
dat_serum_markers_tidy <- dat %>%
  select(c(ID, Responder,Baseline, Eosinophile, CRP, LDH, S100)) %>% 
  gather(serum_marker, value,-c(ID, Responder,Baseline)) %>%
  mutate(log_val = ifelse(is.infinite(log2(value)), 0, log2(value))) %>% 
  filter(!is.na(log_val)) 

# plot 4 markers in separate plots and calculate statistics
plot_serum_markers <- signif_plot_Melanoma(dat_serum_markers_tidy, x="Responder", y="log_val", p.adj = "fdr",
                     plot.type = "dotplot", significance=FALSE, Legend = FALSE, ylab = "log2 serum marker concentration",
                     method ="t.test", p.label="{p.signif}", facet="serum_marker")

# show results of statistical analysis 
plot_serum_markers$stat_test_results
```
![**Fig. 1: log2 serum marker concentration.** Crossbars show mean ± sd. Unequal variances t-test. Raw p-values.](https://raw.githubusercontent.com/MBender1992/MelanomaStudy/Marc/Results/serum%20markers.png)


## 2.2 miRNAs

```{r, message = FALSE}
# tidy miRNA data.....................................................................................................
dat_miRNA_tidy <- dat %>% 
  # only use data where miRNA data was measured and responder status is known
  filter(miRExpAssess == 1 & !is.na(Responder)) %>%
  gather(miRNA, expression, contains("hsa")) %>%
  mutate(miRNA = str_replace_all(.$miRNA, "hsa-","")) %>%
  mutate(log_exp = log2(expression))

# Plot miRNA data
plot_miRNA <- signif_plot_Melanoma(dat_miRNA_tidy, x="Responder", y="log_exp", signif=0.05, p.adj = "fdr", 
                     plot.type = "dotplot", significance=F, Legend = F, var.equal = F,
                     method ="t.test", p.label="p = {round(p,4)}",p.size = 3, facet="miRNA")

# show results of statistical analysis 
plot_miRNA$stat_test_results
```

![**Fig. 2: log2 miRNA expression (a.u.).** Crossbars show mean ± sd. Unequal variances t-test. Raw p-values.](https://raw.githubusercontent.com/MBender1992/MelanomaStudy/Marc/Results/miRNAs.png)





# 3 Modelling process
## 3.1 Data preprocessing and EDA

The initial data was filtered to only keep rows with data for immunotherapy response and miRNA expression resulting in n = 61 for the modelling process. 
Each categorical variable was checked for enough elements in each group when split by immunotherapy response to remove features with high uncertainty (a model using e.g. ECOG containing 2 patients with immunotherapy response and ECOG 2 and 4 patients without immunotherapy and ECOG 2 will be rather uninformative as this variable adds noise instead of useful information). On top of that breslow thickness and BRAF status were dropped from the analysis as the former is highly correlated with the AJCC stage and included in that variable and the latter is highly correlated with prior anti-BRAF treatment (which is a more useful variable as it is known, that patients with a prior BRAF therapy have worse immunotherapy response).

```{r}
#####################################
## 
## 1.Data preprocessing
##
#####################################

# filter data to keep only rows with data for immunotherapy response and miRNA expression and transform categorical variables to factors
dat_fct <- dat %>%
  filter(miRExpAssess == 1 & !is.na(Responder)) %>% # n = 61
  select(-c(TRIM_PDL1_Expression , miRExpAssess, therapy_at_blood_draw)) %>%
  mutate( across(c(Responder, Stadium, Baseline, BRAF, ECOG, subtype, localization,
                   sex, brainMet, adjuvant_IFN, organsInvolved, nras, prior_BRAF_therapy), as.factor)) 

# check each factor for enough elements when split by responder
xtabs(~ Responder + Stadium, data=dat_fct) 
xtabs(~ Responder + BRAF, data=dat_fct)
xtabs(~ Responder + Baseline, data=dat_fct)
xtabs(~ Responder + ECOG, data=dat_fct) # not enough samples in ECOG2
xtabs(~ Responder + subtype, data=dat_fct) # too many groups with few samples
xtabs(~ Responder + localization, data=dat_fct) # too many groups with few samples
xtabs(~ Responder + sex, data=dat_fct)
xtabs(~ Responder + brainMet, data=dat_fct)
xtabs(~ Responder + adjuvant_IFN, data=dat_fct)
xtabs(~ Responder + organsInvolved, data=dat_fct) # too few observations
xtabs(~ Responder + nras, data=dat_fct) # too few observations
xtabs(~ Responder + prior_BRAF_therapy, data=dat_fct)

# remove columns that yield high uncertainty
dat_fct$ECOG <- NULL
dat_fct$subtype <- NULL
dat_fct$localization <- NULL
dat_fct$nras <- NULL
dat_fct$Baseline <- NULL
dat_fct$ID <- NULL
dat_fct$organsInvolved <- NULL
dat_fct$breslow_thickness_mm <- NULL # included in stage variable
dat_fct$BRAF <- NULL # highly correlated with prior BRAF therapy and therefore rather adds noise to the model

```

The dataset contained several variables with missing values. To balance uncertainty introduced by replacing missing values and keeping important variables 
variables with less than 20 % missing values were kept in the analysis and missing values were imputed using a random forest algorithm from the missForest package in R. 

```{r}
#####################################
## 
## 1.a Impute missing values
##
#####################################

# detect percentage of NAs in each column
NAs <- sapply(dat_fct, function(df){
  sum(is.na(df) ==TRUE)/length(df);
})

# remove columns with more than 20 % NAs
dat_fct <- dat_fct[, -which(NAs > 0.2)]

# convert factor columns to numerical 
dat_fct$Stadium <- ifelse(dat_fct$Stadium == "II", 2,ifelse(dat_fct$Stadium == "III", 3, 4))
dat_fct$sex <- ifelse(dat_fct$sex == "m", 1, 0)
dat_fct$brainMet <- ifelse(dat_fct$brainMet == "ja", 1, 0)
dat_fct$prior_BRAF_therapy <- parse_number(as.character(dat_fct$prior_BRAF_therapy))

# impute missing values with random forest algorithm
set.seed(25)
dat_imp <- dat_fct %>% 
  select_if(is.numeric) %>%
  as.data.frame() %>%
  missForest() %>%
  .$ximp %>%
  # replace calculated probabilities by the factor
  mutate(Stadium = round(Stadium), 
         Alter = round(Alter), 
         brainMet = ifelse(brainMet > 0.5, 1,0), 
         prior_BRAF_therapy = ifelse(prior_BRAF_therapy > 0.5, 1, 0))

# replace numerical values by factor for encoding later
dat_imp$sex <- factor(dat_imp$sex, levels = c(0,1), labels = c("w", "m"))
dat_imp$brainMet <- factor(dat_imp$brainMet, levels = c(0,1), labels = c("no", "yes"))
dat_imp$prior_BRAF_therapy  <- factor(dat_imp$prior_BRAF_therapy, levels = c(0,1), labels = c("no", "yes"))

# replacing NAs with imputed values
dat_fct$Stadium <- dat_imp$Stadium
dat_fct$S100 <- dat_imp$S100
dat_fct$Alter<- dat_imp$Alter
dat_fct$brainMet <- dat_imp$brainMet
dat_fct$prior_BRAF_therapy <- dat_imp$prior_BRAF_therapy
dat_fct$sex <- dat_imp$sex
```


To inspect the data and underlying structure the data was split randomly into training and test set in a 70/30 ratio for explorative data analysis. (This split ensured that decisions based on EDA were less than biased than with using the whole dataset).


```{r}

#####################################
##
##  1.b EDA on training set (to avoid drawing conclusions including the test set)
##
#####################################

# define test and training set
set.seed(123)
ind.train <- createDataPartition(dat_fct$Responder, p = 0.7, list = FALSE)

train.EDA  <- dat_fct[ind.train, ] # n = 43
test.EDA <- dat_fct[-ind.train, ] # n = 18



# change data structure for ggplot
dat_miR <- train.EDA %>% 
  select(contains("mir")) %>% 
  gather("miRNA", "expression")

# draw histograms for all miRNAs
miR_hist <- dat_miR %>% 
  ggplot(aes(expression)) +
  geom_histogram(color = "black", fill = "grey") +
  facet_wrap(~miRNA, scales = "free") + 
  theme_bw()

# draw qqplots for all miRNAs
miR_qq <- dat_miR %>% 
  ggplot(aes(sample = expression)) +
  geom_qq() +
  geom_qq_line() +
  facet_wrap(~miRNA, scales = "free")+
  theme_bw()
```
![**Fig. 3: Histogram of miRNA epression** .](https://raw.githubusercontent.com/MBender1992/MelanomaStudy/Marc/Results/miRNA_histogram.png)

![**Fig. 4: qqplot of miRNA expression** .](https://raw.githubusercontent.com/MBender1992/MelanomaStudy/Marc/Results/miRNA_qq.png)

Some samples show heavy tailed distributions (e.g.miR-101, miR-205, miR-9). As expression values follow a log-normal distribution miRNA data has been log-transformed to account for this fact. 


```{r}
# draw histograms for all miRNAs log-transformed
miR_hist_log <- dat_miR %>% 
  ggplot(aes(log(expression))) +
  geom_histogram() +
  facet_wrap(~miRNA, scales = "free")

# draw qqplots for all miRNAs log-transformed
miR_qq_log <- dat_miR %>% 
  ggplot(aes(sample = log(expression))) +
  geom_qq() +
  geom_qq_line() +
  facet_wrap(~miRNA, scales = "free")

## log-transforming miRNA expression improves approximation to normality and gene expression data is known to be log-normal distributed 
## log-transformed miRNA values were used for ML
```

![**Fig. 5: Histogram of log-transformed miRNA expression** .](https://raw.githubusercontent.com/MBender1992/MelanomaStudy/Marc/Results/miRNA_hist_log.png)

![**Fig. 6: qqplot of log-transformed miRNA expression** .](https://raw.githubusercontent.com/MBender1992/MelanomaStudy/Marc/Results/miRNA_qq_log.png)


```{r}
# histogram of numerical variables that are not miRNAs
par(mfrow = c(2,4))
par(mar=c(0.5, 4.5, 0.5, 0.5))

# original expression values
hist(train.data$LDH, main = "LDH")
hist(train.data$Eosinophile, main = "Eosinophile")
hist(train.data$S100, main = "S100")
hist(train.data$CRP, main = "CRP")

# log-transformed expression values
hist(log(train.data$LDH), main = "log-transformed LDH")
hist(log(train.data$Eosinophile), main = "log-transformed Eosinophile")
hist(log(train.data$S100), main = "log-transformed S100")
hist(log(train.data$CRP), main = "log-transformed CRP")
```

Log-transformation improved the approximation of a normal distribution also in serum markers, thus miRNA expression and serum marker levels were log-transformed in the whole dataset.


```{r}
#####################################
##
## 1.c log-transformation 
##
#####################################

# transform the whole dataset
tmp <- dat_fct %>% select(where(is.numeric))
fctrs <- dat_fct %>% select(!where(is.numeric))
dat_log <- data.frame(cbind(log(tmp+1), fctrs)) %>% 
  mutate(Responder = factor(Responder, levels = c("nein", "ja")))
```

## 3.2 Lasso Regression


Guideline: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5217944/

Glm modeling mit glm und Koeffizienten oder glmnet? 

Explanation of Ridge, Lasso and Elastic Net Regularization

relaxed Lasso: Meinshausen 2007

https://machinelearningmastery.com/nested-cross-validation-for-machine-learning-with-python/

Using nested cv as it closely resembles test on independent validation set (https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-7-91)
and leads to less bias than traditional cv: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1397873/

discussion why we used this challenging dataset: Papr: A feature agnostic approach for glaucoma detection in OCT volumes